distribution/lib/Standard/Table/0.2.32-SNAPSHOT/src/Data/Matching.enso

```diff
@@ -0,0 +1,167 @@
+     depends on the matching strategy.
+   - reorder: Specifies whether the matched objects should be reordered according
+     to the order of the matching criteria which matched them.
+     If it is set to False, the matched objects are returned in the same order as
+     in the input.
+     If it is set to True, the matched objects are returned in the order of the
+     criteria which matched them. If a single name has been matched by multiple
+     criteria, it is placed in the group belonging to the first criterion on the
+     list which matched it. If the group of objects matched by a single criterion
+     has more than one element, their relative order is the same as in the
+     input.
+   - name_mapper: A function mapping the provided object to its name which will
+     then be matched with the provided matching criteria. It is set to the
+     identity function by default, thus requiring the input to just be a list of
+     objects to match. But it can be overridden to match more complex objects.
+   - matching_strategy: A Matching_Strategy instance specifying how the
+     criterion should be interpreted.
+   - on_missing: Specifies the behavior when a criterion did not match any
+     objects. By default, a warning is issued, but the operation proceeds.
+     If it is set to `Report_Error`, the operation fails with a dataflow error.
+     If it is set to `Ignore`, the operation proceeds without errors or
+     warnings.
+   - warning_callback: A callback specifying the function to call when issuing a
+     warning. This is a temporary workaround to allow for testing the warning
+     mechanism. Once the proper warning system is implemented, this argument
+     will become obsolete and will be removed. No user code should use this
+     argument, as it will be removed in the future.
+
+   > Example
+     Selects objects matching one of the provided patterns, preserving the input order.
+
+         Matching.match_criteria ["foo", "foobar", "quux", "baz", "Foo"] [".*ba.*", "f.*"] matching_strategy=(Regex case_sensitivity=True) == ["foo", "foobar", "baz"]
+
+   > Example
+     Selects pairs matching their first element with the provided criteria and
+     ordering the result according to the order of criteria that matched them.
+
+         Matching.match_criteria [Pair "foo" 42, Pair "bar" 33, Pair "baz" 10, Pair "foo" 0, Pair 10 10] ["bar", "foo"] reorder=True name_mapper=_.name == [Pair "bar" 33, Pair "foo" 42, Pair "foo" 0]
+match_criteria : Vector Any -> Vector Text -> Boolean -> (Any -> Text) -> Matching_Strategy -> Problem_Behavior -> (Text -> Nothing) -> Vector Any ! No_Matches_Found
+match_criteria objects criteria reorder=False name_mapper=(x->x) matching_strategy=(Exact case_sensitivity=True) on_missing=Report_Warning warning_callback=here.report_warning_placeholder = Panic.recover <|
+    [objects, criteria, reorder, name_mapper, matching_strategy, on_missing, warning_callback] . each Panic.rethrow
```

Comment:
Why do we need to re throw panics?

Reply:
We felt that if `[].match An_Error` should result in the panic being rethrown.

This felt true for all the arguments in this function, that the errors should be propagated along the dataflow

Reply:
`Panic.rethrow` takes any dataflow error and raises it to a Panic. Then, we have a `Panic.catch` in the outer layer of the function which again 'downgrades' it to a dataflow error.

This is done to satisfy the error handling unit tests which can be found in this PR. Without that, some errors on the inputs can be silently ignored if the other input is an empty list - and we decided that this is bad (why a list being empty or not should change whether an error from a different argument is forwarded or ignored?).

Reply:
I don't like the idea of converting one errors to others and back – this has some performanc cost for sure + the design is strange, especially if the requirements from the testing framework influences the design of the code. Is there any other/better way of doing that? CC @radeusgd @kustosz

Reply:
This approach with `Panic.rethrow` has already been used in some places, like the Database library and in even more places we use `Panic.catch`+`Panic.throw` to ensure that the error is not lost - the issue is that in some implementations the result does not depend (in a data-flow way) on some places which can raise errors and this approach ensures that we do not lose it.

(The lack of dependency happens mostly if an input list is empty so the actual "loops" which traverse that list and touch these other arguments never really run so they do not introduce this dependency)

But indeed, at least in this case, we could change it to just introduce a dataflow dependency somehow - although I'm not sure how we should do that, we'd need some function like `seq : a ! e1 -> b ! e2 -> a ! (e1 | e2)` which ensures that errors 
from `b` are promoted to the result even if it does not depend on them. I guess it can be implemented by:
```
seq a b = case b of _ -> a
```

Reply:
Regarding performance, I don't think there's a problem – an `instanceof` check is cheap, and then it's just throwing an exception to abort early (at which point performance is secondary, we're escaping execution anyway).

Regarding the cleanliness of such design – I tend to use it for imperative methods, i.e. ones where there's no dataflow connection between the inputs and the outputs. In this case we want to still see the warnings, but they are dataflow, so no luck. I don't think it's a good idea to do it for a nice, dataflow-y method. There was an error, but its value was unused, therefore things are fine. Kind of like going `const 1 undefined` in Haskell. But that's a decision for the library designers to make. My take would be: what's the point of using dataflow errors, if you don't care about their dataflow-ness...

Reply:
### Implementation Code Style

I tihnk we are generally trying to figure out the right code style - so I understand that performance-wise both should be fine, but for dataflow-y methods it is probably better to keep a single style instead of adding these imperative Panics.

btw. if I write a `seq` using an implementation like above, or maybe `seq ~a b = case b of _ -> a`, I can't still avoid
```
foo = may have errors
bar = does an expensive computation
result = seq bar foo
```
to avoid computing `bar` if it is thrown away due to the dataflow error - right? So in this case Panics will actually have better/more predictable performance.

### User-facing API style

As for the user-facing interface - I understand your argument regarding `cons` and I guess for some methods it does indeed make sense. But IMO it can be surprising if you do `Matching.match_criteria [] [] (Error.throw "error computing if I wanted to reorder or not")` and not get any errors, but if the lists are not empty anymore you will have them. Well, there are advantages and disadvantages:
- it is nice that the computation kind-of opportunistically avoids the error because it "knows" its result is, in this particular case, independent of this argument;
- on the other hand, IMO it may be problematic as the user can easily lose track of errors - after all, it may be a problem if computing some of these settings resulted in an error, but we silently ignore it because another input was empty;
- and regarding the above - the user may prepare a workflow and think that there are no problems - but then the input changes and the error starts being a problem.

Of course, regarding the last point - it will always be the case because we can have stuff like `list.map x-> if x < 100 then x + 1 else Error.throw "Hmmm..."` and here also changing the input will either make us have errors or not. So it is unavoidable in general.

Regarding the 2nd point - maybe this is not a big issue if we have a pane showing all errors in a given workflow - then the user should still notice that some function computing these parameters failed even though they turned out to be not needed and the error did not propagate. But if this computation is inside of some sub-graph, then if the error does not propagate, it will not be visible outside.

cc: @jdunkerley @NedHarding @wdanilo I think you may want to have a look at this second part, as it influences how we will be handling errors in our user-facing components.

Reply:
We will probably proceed with this PR assuming what we have currently is good.

But we still need to clarify this design-wise for this and any future APIs - if the decision is changed than what we assumed for now, we can very easily update this in a separate PR (just removing/changing the rethrows).